inst/include/dust/random/normal_ziggurat.hpp

```diff
@@ -0,0 +1,75 @@
+#ifndef DUST_RANDOM_NORMAL_ZIGGURAT_HPP
+#define DUST_RANDOM_NORMAL_ZIGGURAT_HPP
+
+#include <cmath>
+
+#include "dust/random/generator.hpp"
+#include "dust/random/normal_ziggurat_tables.hpp"
+
+namespace dust {
+namespace random {
+
+__nv_exec_check_disable__
+namespace {
+template <typename real_type, typename rng_state_type>
+__host__ __device__
+real_type normal_ziggurat_tail(rng_state_type& rng_state, real_type x1,
+                               bool negative) {
+  real_type ret;
+  do {
+    const auto u1 = random_real<real_type>(rng_state);
+    const auto u2 = random_real<real_type>(rng_state);
+    const auto x = std::log(u1) / x1;
+    const auto y = std::log(u2);
+    if (- 2 * y > x * x) {
+      ret = negative ? x - x1 : x1 - x;
+      break;
+    }
+  } while (true);
+  return ret;
+}
+}
+
+// TODO: this will not work efficiently for float types because we
+// don't have float tables for 'x' and 'y'; getting them is not easy
+// without requiring c++14 either. The lower loop using 'x' could
```

Comment:
How come we are able to manage this with binomial draws?

Reply:
The binomial algorithm doesn't need to use the random numbers; the tables are only used in `stirling_approx_tail`, so we have fully specialised templates on `real_type` (and some ugly names `k_tail_values_d` and `stirling_approx_tail_f`). With C++14 we can make these template variables which removes the naming problem but because the algorithm here needs to use one or two random numbers along side the constants we'd end up with trying to make a partially specialised template (providing real_type but leaving rng_state_type open).

There are some solutions, but I'd like to try and implement these separately and compare timings to make sure that we don't end up paying too much on the CPU case